### ISSUE TITLE
Add `foo(Foo.Builder foo)` to builders

### ISSUE THREAD

**Issue by ukslim** [immutables/immutables#391]
In Freeman/Pryce *Growing Object Oriented Software Guided by Tests*, it's suggested that builder methods should take builders as parameters.

So instead of:

      Order order = Order.builder()
              .withAddress(Address.builder().postcode("SY24 4BG").build())
              .build();

We can have:

      Order order = Order.builder()
              .withAddress(Address.builder().postcode("SY24 4BG"))
              .build();

For complex objects, you can make things quite a lot neater this way.

Could Immutables support this style (in addition to what it already has)?

**Comment by elucash** [comment]
It's seems like a good idea, but is it worth it? `.build()`: 8 characters in exchange of additional method overload for each such attribute plus additional code to analyze. We are currently developing special "encoding" functionality and best we can do with this is to try to make possible to be able to add such overloads with extension. Thank you!

**Comment by cdeszaq** [comment]
This sounds like the sort of thing a style could address, since it seems like a configuration flag as to how the code should be generated? We have some use-cases where we have a very deep tree of immutables, with higher-level immutables having to call `.build()` upwards of 30 times. It may be only 8 characters if you have to do this once, but the clutter adds up quickly!

**Comment by markhobson** [comment]
I agree that a new style would be a good way to address this. In the meantime this can be achieved by manually customising the builder:

```
public class Main {
    
    @Value.Immutable
    public interface Order {
        
        abstract class Builder {
            abstract ImmutableOrder.Builder address(Address address);
            
            public ImmutableOrder.Builder address(ImmutableAddress.Builder addressBuilder) {
                return address(addressBuilder.build());
            }
        }
        
        Address getAddress();
        
        static ImmutableOrder.Builder builder() {
            return ImmutableOrder.builder();
        }
    }
    
    @Value.Immutable
    public interface Address {
        
        String getPostcode();
        
        static ImmutableAddress.Builder builder() {
            return ImmutableAddress.builder();
        }
    }
    
    public static void main(String[] args) {
        Order order1 = Order.builder()
            .address(Address.builder().postcode("SY24 4BG").build())
            .build();
        
        Order order2 = Order.builder()
            .address(Address.builder().postcode("SY24 4BG"))
            .build();
    }
}
```
